### PR TITLE
Avoid local cleanup of vmi until vmi is in a finalized state

### PR DESCRIPTION
**Pull request by davidvossel**
Under certain failure conditions, it's possible for the domain to never hit a running state and go directly to shutdown. This occurred while I was performing a PoC of a new feature where container disks can be a backing disk for a qcow2 image on a PVC. The disk on the PVC wasn't writeable, which caused the domain to immediately go to a failed "shutdown" state.

The result of this is that the VMI remained forever as "scheduled". It never moved to Running and was never finalized. Further more the VMI can't be deleted because the finalizer is only removed once the VMI reaches a finalized state. To make matters worse, if the VMI is owned by a VM, the VM can't restart because the VMI will never delete. 

This was ultimately caused by logic that assumes that if a domain doesn't exist and the vmi is being deleted, that the vmi must have already reached a Finalized state. This isn't a safe assumption, as my situation proved.

The fix is to never perform local cleanup unless the VMI is finalized. As long as the command sockets and other local data exists, the VMI will definitely timeout and reach a finalized state eventually. Cleaning up before the VMI is finalized negates that though.

```release-note
Fixes rare situation where vmi may not properly terminate if failure occurs before domain starts.
```

### CODE REVIEW

File: pkg/virt-handler/vm_test.go [hunk 1]
```diff
@@ -442,6 +442,35 @@ var _ = Describe("VirtualMachineInstance", func() {
...
+			vmiFeeder.Add(vmi)
+
+			client.EXPECT().SyncVirtualMachine(vmi, gomock.Any())
+
+			controller.Execute()
+			Expect(mockQueue.Len()).To(Equal(0))
+			Expect(mockQueue.GetRateLimitedEnqueueCount()).To(Equal(0))
+			_, err := os.Stat(mockWatchdog.File(vmi))
+			Expect(os.IsNotExist(err)).To(BeFalse())
+			Expect(len(controller.phase1NetworkSetupCache)).To(Equal(1))
```
Comment (danielBelenky) [inline pkg/virt-handler/vm_test.go:471]: Will it make sense to also check here that the domain doesn't exist in the domain store anymore?

Comment (davidvossel) [reply pkg/virt-handler/vm_test.go:471]: there is no domain in this test. it was never added to the store. a domain is only added to the store by the test case itself here.

**Comment by davidvossel** [comment]
> Also, should we add a functional test for this case?

At this point it's unclear how to reliably reproduce this in a functional test.

**Pull request merged**